contracts/sui/xcall/sources/main.move

```
@@ -0,0 +1,547 @@
+                    is_response,
+                    ctx);
+
+                i=i+1;
+            };
+
+            fee
+    }
+
+    fun get_next_sequence(self:&mut Storage):u128 {
+        let sn=xcall_state::get_next_sequence(self);
+        sn
+    }
+
+
+    fun get_next_req_id(self:&mut Storage):u128 {
+        let req_id=xcall_state::get_next_request_id(self);
+        req_id
+    }
+
+    entry fun set_protocol_fee(self:&mut Storage,admin:&AdminCap,fee:u64){
+        self.enforce_version(CURRENT_VERSION);
+        xcall_state::set_protocol_fee(self,fee);
+    }
+
+    entry fun set_protocol_fee_handler(self:&mut Storage,admin:&AdminCap,fee_handler:address){
+        self.enforce_version(CURRENT_VERSION);
+        xcall_state::set_protocol_fee_handler(self,fee_handler);
+    }
+
+    public fun send_call(self:&mut Storage,fee: Coin<SUI>,idCap:&IDCap,to:String,envelope_bytes:vector<u8>,ctx: &mut TxContext){
+        let envelope=envelope::decode(&envelope_bytes);
+        let to = network_address::from_string(to);
+        let from= network_address::create(xcall_state::get_net_id(self),utils::id_to_hex_string(&object::id(idCap)));
+
+        let remaining = send_call_inner(self,fee,from,to,envelope,ctx);
+        transfer::public_transfer(remaining, ctx.sender());
+    }
+   
+
+    public fun handle_message(self:&mut Storage,
+    cap:&ConnCap, 
+    from:String, 
+    msg:vector<u8>,
+    ctx: &mut TxContext){
+        self.enforce_version(CURRENT_VERSION);
+        assert!(from != xcall_state::get_net_id(self),EInvalidNID);
+        let cs_msg = cs_message::decode(&msg);
+        let msg_type = cs_message::msg_type(&cs_msg);
+        let payload = cs_message::payload(&cs_msg);
+
+        if (msg_type == cs_message::request_code()) {
+            handle_request(self,cap,from, payload, ctx);
+        } else if (msg_type == cs_message::result_code()) {
+           
+            handle_result(self, cap, payload, ctx);
+        } else {
+        }
```

We should probably raise a error here if possible